---
title: "Pvalues Generator"
author: "Katriona Goldmann"
output: rmarkdown::html_document
---

```{r setup, include = FALSE, echo = FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      warning = FALSE, 
                      message = FALSE, 
                      fig.height = 7, 
                      fig.width=7, 
                      fig.align = "center")
library(knitr)
library(kableExtra)
source("https://gist.githubusercontent.com/KatrionaGoldmann/3d6dfd6aa4cc5c3940bb72dc49beae02/raw/26e78c78a7a9d096d695c708e8a45830a1d1121a/render_toc.R")
```


<img src="../logo.png" align="right" alt="" width="200" />

```{r toc, echo = FALSE}
render_toc("./pvalues_generator.rmd")
```

---

# Introduction 

This vignette provides an example for creating pvalues objects for the 
volcano3D pipeline using DESeq2 and limma. 

This example consists of a case study from the PEAC 
rheumatoid arthritis project (Pathobiology of Early Arthritis Cohort). 
The methodology has been published in 
['Lewis, Myles J., et al. "Molecular portraits of early rheumatoid arthritis 
identify clinical and treatment response phenotypes." Cell reports 28.9 (2019): 
2455-2470.'
(DOI: 10.1016/j.celrep.2019.07.091)](https://doi.org/10.1016/j.celrep.2019.07.091)
with an interactive web tool available at 
[https://peac.hpc.qmul.ac.uk](https://peac.hpc.qmul.ac.uk).  

## Getting Started

### Install from CRAN

*Not yet publicly available:*

```{r, eval = FALSE}
install.packages("volcano3D")
```

### Install from Github

```{r, eval = FALSE}
library(devtools)
install_github("KatrionaGoldmann/volcano3D")
```

```{r}
library(volcano3D)
```

The sample data used in this vignette can be loaded from the 
[volcano3Ddata package](https://github.com/KatrionaGoldmann/volcano3Ddata). 
This is only possible after volcano3D is imported. 

```{r, eval=FALSE}
install.packages("volcano3Ddata")
```

---

To create a pvalues data frame we require:

1. the txi or expression data with columns representing different samples and 
rows representing different variables
2. The sample data which contains information for each sample in rows. 

```{r}
library(volcano3Ddata)
data("syn_data")
```

```{r, eval=FALSE}
data("syn_txi")
```

```{r, echo=FALSE}
load("~/Downloads/syn_txi.rdata")
```

Check the alignment. 

```{r}
if(! identical(syn_metadata$ID, colnames(syn_txi$abundance))){
    print("mis-aligned")
}
```

Make sure there are only three possible contrasts (i.e. three unique Pathotypes)

```{r}
if(length(levels(syn_metadata$Pathotype)) != 3){
    print("The number of unique pathotypes must qual 3")
}
```

Comparisons of interest:

```{r, results="markup"}
groups <- levels(syn_metadata$Pathotype)
contrasts <- c(paste(groups[1], groups[2], sep="-"), 
               paste(groups[2], groups[3], sep="-"), 
               paste(groups[3], groups[1], sep="-"))

print(paste(contrasts, collapse=", "))
```

```{r, echo=FALSE}
load("~/Desktop/Pvalues.rdata")
```

# DESeq method

Use the DESeq package to calculate the differential expression between groups. 
The vignette for this package can be found [here](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)

```{r, eval=FALSE}
library(DESeq2)
dds = DESeqDataSetFromTximport(txi = syn_txi, 
                               colData = syn_metadata, 
                               design = ~Pathotype+Batch+Gender)

dds_DE <- DESeq(dds)
dds_LRT <- DESeq(dds, test = "LRT", reduced = ~Batch+Gender, parallel = TRUE) 
```

Output the results into a dataframe. First for each contrast:

```{r, eval=FALSE}
Pvals_DESeq_DE <- lapply(contrasts, function(x){
    vars <- unlist(strsplit(x, split="-"))
    out <- results(dds_DE, contrast=c("Pathotype", vars))
    out <- out[match(rownames(syn_txi$counts), rownames(out)), ]
    out <- out[,c("pvalue", "padj", "log2FoldChange")]
})
```

Then for a multi-group test. In this case we use the Likelihood ratio test. 

```{r, eval=FALSE}
LRT <- results(dds_LRT, parallel=TRUE)[, c("pvalue", "padj", "log2FoldChange")]
LRT <- LRT[match(rownames(syn_txi$counts), rownames(LRT)), ]
```

Combine these into one pvalues object: 

```{r, eval=FALSE}
Pvals_DESeq <- cbind(Pvals_DESeq_DE[[1]], 
                     Pvals_DESeq_DE[[2]], 
                     Pvals_DESeq_DE[[3]], 
                     LRT)
colnames(Pvals_DESeq) <- c(paste(rep(contrasts, each=3), 
                           rep(colnames(Pvals_DESeq_DE[[1]]), 3)), 
                     paste("LRT", colnames(LRT)))

Pvals_DESeq <- data.frame(Pvals_DESeq[complete.cases(Pvals_DESeq), ], 
                          check.names = FALSE)
```

Now we can inspect:

```{r, eval=FALSE}
head(Pvals_DESeq)
```

```{r, echo=FALSE}
head(Pvals_DESeq) %>%
    kable() %>%
    kable_styling(font_size=8.7)
```

<!-- # Limma method -->

<!-- Use the Limma and edgeR package to calculate the differential expression  -->
<!-- between groups. The vignette for this package can be found [here](https://www.bioconductor.org/packages/devel/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html) -->

<!-- ```{r} -->
<!-- library(limma) -->
<!-- library(edgeR) -->
<!-- load("~/Downloads/syn_txi.rdata") -->
<!-- ``` -->

<!-- Define the design for Limma -->

<!-- ```{r, eval=FALSE} -->
<!-- design.df <- model.matrix(~ 0 + Pathotype + Batch + Gender, data=syn_metadata)  -->
<!-- colnames(design.df) <- gsub("[[:punct:]]|factor|syn_metadata|Pathotype|Batch",  -->
<!--                             "", colnames(design.df)) -->

<!-- contrast.matrix <- makeContrasts(contrasts=c(contrasts, "B", "GenderB"),  -->
<!--                                  levels = c(groups,"B", "GenderB") ) -->
<!-- ``` -->

<!-- Remove the batch effect: -->

<!-- ```{r} -->
<!-- exp <- data.frame(apply(syn_rld, 2, as.numeric)) -->
<!-- rownames(exp) <- make.names(rownames(syn_rld), unique=T) -->
<!-- ``` -->

<!-- Calculate the differential expression -->

<!-- ```{r, eval=FALSE} -->
<!-- fit <- lmFit(exp, design.df) -->
<!-- fit2  <- contrasts.fit(fit, contrasts=contrast.matrix) -->
<!-- fit3  <- eBayes(fit2, trend = TRUE) -->
<!-- ``` -->

<!-- Output the p-values first for individual comparisons: -->

<!-- ```{r, eval=FALSE} -->
<!-- Pvals_limma_DE <- lapply(contrasts, function(x){ -->
<!--     out <- topTable(fit3,  -->
<!--              coef=which(colnames(fit3$coefficients) == x),  -->
<!--              number=Inf,  -->
<!--              sort.by="P")[,c("P.Value", "adj.P.Val", "logFC")] -->
<!--     out <- out[match(rownames(exp), rownames(out)), ] -->
<!--     colnames(out) <- c("pvalue", "padj", "log2FoldChange") -->
<!--     out -->
<!-- }) -->
<!-- ``` -->

<!-- Then for an empirical Bayes multi-group test: -->

<!-- ```{r, eval=FALSE} -->
<!-- eB <- topTable(fit3, number=Inf,  -->
<!--                 coef=which(colnames(fit3$coefficients) %in% contrasts)) -->
<!-- eB <- eB[match(rownames(exp), rownames(eB)), ] -->
<!-- eB <- eB[, ! colnames(eB) %in% c("AveExpr", "F")] -->
<!-- colnames(eB) = c(contrasts,  "pvalue", "padj" ) -->
<!-- ``` -->

<!-- Combined into one object -->

<!-- ```{r, eval=FALSE} -->
<!-- Pvals_limma <- cbind(Pvals_limma_DE[[1]],  -->
<!--                      Pvals_limma_DE[[2]],  -->
<!--                      Pvals_limma_DE[[3]],  -->
<!--                      "eBayes" = eB) -->
<!-- colnames(Pvals_limma) <- c(paste(rep(contrasts, each=3),  -->
<!--                            rep(colnames(Pvals_limma_DE[[1]]), 3)),  -->
<!--                      paste("eBayes", colnames(eB))) -->

<!-- Pvals_limma <- data.frame(Pvals_limma[complete.cases(Pvals_limma), ],  -->
<!--                           check.names = FALSE) -->
<!-- ``` -->

<!-- Now we can inspect: -->

<!-- ```{r, eval=FALSE} -->
<!-- head(Pvals_limma) -->
<!-- ``` -->

<!-- ```{r, echo=FALSE} -->
<!-- Pvals_limma[, 1:(ncol(Pvals_limma)-1)] = format( -->
<!--     Pvals_limma[, 1:(ncol(Pvals_limma)-1)], digits = 3) -->

<!-- head(Pvals_limma) %>% -->
<!--     kable() %>% -->
<!--     kable_styling(font_size=8.7) -->
<!-- ``` -->

<!-- These objects can be used as pvalues dataframes in downstream analysis with the -->
<!-- volcano3D package.  -->
