---
title: "volcano3D Vignette"
author: "Katriona Goldmann"
output: rmarkdown::html_document
vignette: >
%\VignetteIndexEntry{volcano3D package}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
  ---
  
  ```{r setup, include = FALSE, echo = FALSE}
  knitr::opts_chunk$set(echo = TRUE, 
  warning = FALSE, 
  message = FALSE, 
  fig.height = 7, 
  fig.width=7, 
  fig.align = "center")
  library(knitr)
  library(kableExtra)
  source("https://gist.githubusercontent.com/KatrionaGoldmann/3d6dfd6aa4cc5c3940bb72dc49beae02/raw/26e78c78a7a9d096d695c708e8a45830a1d1121a/render_toc.R")
  ```
  
  
  <img src="../logo.png" align="right" alt="" width="200" />
  
  ```{r toc, echo = FALSE}
  render_toc("./Vignette.rmd")
  ```
  
  ---
  
  [![HitCount](http://hits.dwyl.com/KatrionaGoldmann/volcano3D.svg)](http://hits.dwyl.com/KatrionaGoldmann/volcano3D)
  [![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
  [![CRAN status](https://www.r-pkg.org/badges/version/volcano3D)](https://cran.r-project.org/package=volcano3D)
  [![Codecov test coverage](https://codecov.io/gh/r-lib/volcano3D/branch/master/graph/badge.svg)](https://codecov.io/gh/r-lib/volcano3D?branch=master)
  [![Downloads](https://cranlogs.r-pkg.org/badges/volcano3D?color=blue)](https://cran.rstudio.com/package=volcano3D)
  
  # Introduction 
  
  The volcano3D package enables exploration of probes differentially 
  expressed between three groups. Its main purpose is for the
  visualisation of differentially expressed genes in a three-dimensional
  volcano plot. These plots can be converted to interactive visualisations using
  plotly. 
  
  This vignette will consist of a number of case studies focusing on the PEAC 
  rheumatoid arthritis study (Pathobiology of Early Arthritis Cohort). 
  The methodology has been published in 
  ['Lewis, Myles J., et al. "Molecular portraits of early rheumatoid arthritis 
  identify clinical and treatment response phenotypes." Cell reports 28.9 (2019): 
  2455-2470.'
  (DOI: 10.1016/j.celrep.2019.07.091)](https://doi.org/10.1016/j.celrep.2019.07.091)
  with an interactive web tool available at 
  [https://peac.hpc.qmul.ac.uk](https://peac.hpc.qmul.ac.uk).  
  
  ## Getting Started
  
  ### Prerequisites
  
  * [ggplot2](https://cran.r-project.org/web/packages/ggplot2/index.html)
  * [ggpubr](https://cran.r-project.org/web/packages/ggpubr/index.html)
  * [ggrepel](https://cran.r-project.org/web/packages/ggrepel/index.html)
  * [plotly](https://cran.r-project.org/web/packages/plotly/index.html)
  
  ### Install from CRAN
  
  *Not yet publicly available:*
  
  ```{r, eval = FALSE}
  install.packages("volcano3D")
```


### Install from Github

```{r, eval = FALSE}
library(devtools)
install_github("KatrionaGoldmann/volcano3D")
```

```{r, eval = FALSE}
library(volcano3D)
```

```{r, echo=FALSE}
# Secretly load in the local files
sapply(list.files("./R/"), source)
```


### Sample Data

The sample data can then also be installed (this can only be done 
after volcano3D is imported)

```{r, eval=FALSE}
install.packages("volcano3Ddata")
```

### volcano3D data

The PEAC data for this package is automatically loaded from the volcano3Ddata 
package repo ([source](https://github.com/KatrionaGoldmann/volcano3Ddata)). 

---

# Dictionary 

Variables used in this vignette:


| <span style="color:maroon">Variable</span> | <span style="color:maroon">Definition </span>  |
|:-----------------:|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| <b>contrast</b> | the variable by which samples can be split into three groups. |
| <b>groups</b> | the three levels/categories of the contrast variable. |
| <b>comparison</b> | two groups between which a statistical test can be performed. There are three comparisons total. For the examples outlined in this vignette we look at comparisons: 'lymphoid-myeloid', 'lymphoid-fibroid' and 'myeloid-fibroid'. |
| <b>p</b> | p value |
| <b>FC</b> | fold change |
| <b>padj</b> | adjusted p value |
| <b>suffix</b> | the tail word in a column name. In this package it states the statistical parameter (e.g. _logFC is the log FC variable)  |
| <b>prefix</b> | the leading word in a column name. In this package it states the statistical test (e.g. LRT is the likelihood ratio test). |
| <b>polar</b> | A polar coordinates object, of S4 class, containing the expression data, sample data, pvalues and polar coordinates. |

---

# Examples

This vignette will demonstrate the power of this package using a basic example
from the [PEAC data set](http://www.peac-mrc.mds.qmul.ac.uk). This analysis is
separated by biopsy/tissue type: blood and synovium. Here we will focus on the
synovial data. This is stored in the Git repo
[volcano3Ddata](https://github.com/KatrionaGoldmann/volcano3Ddata).

## 1. Synovial Gene Data

First we will set up a differential expression pvalues object (dep) containing:

1. the expression data with columns representing different samples and rows
representing different variables
2. the sample data which contains information for each sample in rows.
3. the pvalues which contains the statistical significance of probes between
groups.

### Creating polar coordinate objects

To create the polar coordinates you must provide the 'polar_coords' function 
with:

* `sampledata`: A data frame containing the sample information. This must
contain:
* an `ID` column: Containing the sample IDs. This must be titled 'ID'.
* a `contrast` column: A column containing the three-level factor used for
contrasts.
* `contrast`: The column name in `sampledata` which contains the three-level
factor used for contrast.
* `pvalues`: A data frame containing:
* three pvalue columns: one for each comparison between groups (column names
of format `paste0(groups[i], "-", groups[j], " ", p_col_suffix)`). We
recommend using [limma](https://www.bioconductor.org/packages/devel/workflows/
vignettes/RNAseq123/inst/doc/limmaWorkflow.html) or
[DESeq](http://bioconductor.org/packages/release/bioc/vignettes/DESeq/inst/doc/
DESeq.pdf) pipelines to calculate these pvalues for gene expression.
* _optional_ fold change columns one for each comparison between groups 
(column namesof format `paste0(groups[i], "-", groups[j], " ", 
fc_col_suffix)`)
* _optional_ adjusted pvalue columns one for each comparison between groups 
(column namesof format `paste0(groups[i], "-", groups[j], " ", 
padj_col_suffix)`)
* an _optional_ multi-group pvalue column: from a multi-group test (column
names of form `paste(multi_group_prefix, p_col_suffix)`). This is typically
generated using ANOVA or likelihood ratio tests between all three groups.
* three fold change columns: one for each comparison between groups
(column names of form `paste0(groups[i], "-", groups[j], " ", fcColSuffix)`)
* an _optional_ multi-group fold change column: from a multi-group test
(column names of form `paste(multi_group_prefix, fc_col_suffix)`). This is
typically generated using ANOVA or likelihood ratio tests comparing all
three groups.
* three _optional_ adjusted pvalue columns: one for each comparison between
groups (column names of form `paste0(groups[i], "-", groups[j], " ",
padj_col_suffix)`)
* an _optional_ multi-group adjusted pvalue column: from a multi-group test
(column names of form `paste(multiGroupPrefix, padjColSuffix)`). This is
typically generated using ANOVA or likelihood ratio tests comparing all
three groups.

with optional:

* `groups`: The groups to be compared (in order). If NULL this defaults to
`levels(sampledata[, 'contrasts'])`.
* `p_col_suffix`: The suffix of column names with pvalues (default is 'pvalue').
* `padj_col_suffix`: The suffix of column names with adjusted pvalues (default
is 'padj'). If NULL the adjusted pvalue is calculated using `p_col_suffix` and
pvalue_method.
* `padjust_method`: The method to calculate adjusted pvalues if not already
provided. Must be one of c('holm', 'hochberg', 'hommel', 'bonferroni', 'BH',
'BY', 'fdr', 'none'). Default is 'BH'.
* `fc_col_suffix`: The suffix of column names with log(fold change) values
(default is 'logFC').
* `multi_group_prefix`: The prefix for columns containing statistics for a
multi-group test (this is typically a likelihood ratio test or ANOVA). Default
is NULL.

Using the synovial biopsies from PEAC we can create a dep object for
differentially expressed genes. First we will install the volcano3Ddata package
if necessary:

```{r}
# Install the data package if required
if(! "volcano3Ddata" %in% installed.packages()[,"Package"]){
  install.packages("volcano3Ddata")
}
```

Then the data can be loaded:

```{r}
library(volcano3Ddata)
data("syn_data")

syn_polar <- polar_coords(sampledata = syn_metadata,
                          contrast = "Pathotype",
                          pvalues = syn_pvalues, 
                          expression = syn_rld, 
                          p_col_suffix = "pvalue", 
                          padj_col_suffix = "padj", 
                          fc_col_suffix = "log2FoldChange", 
                          multi_group_prefix = "LRT", 
                          non_sig_name = "Not Significant", 
                          significance_cutoff = 0.01, 
                          fc_cutoff = 0.3)
```

The pvalues slot should have three statistics for each comparison: pvalue,
adjusted pvalue and logarithmic fold change:

```{r, eval=FALSE}
head(syn_polar@pvalues) 
```

```{r, echo=FALSE}
head(syn_polar@pvalues) %>%
  kable() %>%
  kable_styling(font_size=8.7)
```

The `sig` column in `syn_polar@polar` allows us to determine relative
differences in expression between pathotypes. The '+'
indicates which pathotypes are significantly 'up' compared to others. For
example:

* genes labelled 'Lymphoid+' are significantly up in Lymphoid vs
Myeloid **and** Lymphoid vs Fibroid. Genes such as these which are up in only
one pathotype are coloured
according to the `primary_colours` variable in `polar_coords`.
* genes up in two pathotypes such as 'Lymphoid+Myeloid+' are up in both
Lymphoid **and** Myeloid, therefore Lymphoid vs
Fibroid and Myeloid vs Fibroid are statistically significant. These genes are
coloured according to the `secondary_colours`.
* genes which show now
significant difference between pathotypes (The 'Not Significant' genes) are
coloured `not_sig_colour`.

```{r}
table(syn_polar@polar$sig) %>%
  kable(col.names = c("Significance", "Frequency")) %>%
  kable_styling(full_width = F)
```

### Volcano Plots

If there is a fold change column previously provided, we can now investigate 
the comparisons between pathotypes using the
volcano_trio function. This creates three
[ggplot](https://ggplot2.tidyverse.org/reference/ggplot.html) outputs.

```{r, message = FALSE, fig.height = 2.8}
syn_plots <- volcano_trio(polar = syn_polar,
                          sig_names = c("not significant","significant",
                                        "not significant","significant"),
                          colours = rep(c("grey60",  "salmon"), 2),
                          text_size = 9,
                          shared_legend_size = 0.9,
                          label_rows = c("SLAMF6", "PARP16", "ITM2C"),
                          fc_line = FALSE)

syn_plots$All
```

### Radial Plots

These can be output to an interactive radar plot using radial_plotly. The
`labelRows` variable allows any markers of interest to be labelled.

```{r, warning = FALSE, message = FALSE, out.height = "100%"}
radial_plotly(polar = syn_polar,
              fc_cutoff = 0.1,
              label_rows = c("SLAMF6", "PARP16", "ITM2C"))
```

By hovering over certain point you can also determine genes for
future interrogation.

Similarly we can create a static ggplot image using radial_ggplot:

```{r, fig.height=5}
radial_ggplot(polar = syn_polar,
              fc_cutoff = 0.1,
              label_rows = c("SLAMF6", "PARP16", "ITM2C"),
              marker_size = 2.3,
              legend_size = 10)
```

Alternatively a continuous colour scale can be produced by converting the angle
to a hsv variable. This angle can be offset by `continuous_shift`.

```{r}
radial_ggplot(polar = syn_polar,
              fc_cutoff = 0.1,
              label_rows = c("SLAMF6", "PARP16", "ITM2C"),
              marker_size = 2.3,
              colour_scale = "continuous",
              continuous_shift=120, 
              legend_size = 10)
```


### Boxplots

We can then interrogate any one specific variable as a boxplot, to investigate
these differences. This is build using ggplot2 so can easily be edited by the
user to add features.

```{r, fig.height = 2.8}
plot1 <- boxplot_trio(syn_polar,
                      value = "SLAMF6",
                      test = "wilcox.test",
                      levels_order = c("Lymphoid", "Myeloid", "Fibroid"),
                      box_colours = c("blue", "red", "green3"))

plot2 <- boxplot_trio(syn_polar,
                      value = "PARP16",
                      test = "wilcox.test",
                      levels_order = c("Myeloid", "Fibroid"),
                      box_colours = c("red", "green3"))

plot3 <- boxplot_trio(syn_polar,
                      value = "ITM2C",
                      test = "wilcox.test",
                      levels_order = c("Lymphoid", "Myeloid", "Fibroid"),
                      my_comparisons = list(c("Lymphoid", "Myeloid"),
                                            c("Myeloid", "Fibroid")),
                      box_colours = c("blue", "red", "green3"))

ggarrange(plot1, plot2, plot3, ncol = 3)
```

### Three Dimensional Volcano Plots

The final thing we can look at is the 3D volcano plot which projects
differential gene expression onto cylindrical coordinates.

```{r}
p <- volcano3D(syn_polar,
               label_rows = c("SLAMF6", "PARP16", "ITM2C"),
               label_size = 10,
               xy_aspectratio = 1,
               z_aspectratio = 0.9)

p %>% layout(legend = list(x = 100, y = 0.5))
```

Again this produces an interactive plot. If you have
the orca command-line utility installed, this can be used to
save static images. To install follow the instructions
[here](https://github.com/plotly/orca#installation).

```{r, eval = FALSE}
orca(p, "./volcano_3d_synovium.svg", format = "svg")
```

### Altering the Grid

By default volcano3D generates a grid using 12 spokes. If you wish to override
any of these variables it is possible to pass in your own grid. This is also 
possible for the 3D radial plots where the z elements can be left as NULL. 

By manually creating a grid object it is possible to change the tick points on 
axes as well as the number of radial spokes (n_spokes).  

The default ticks are calculated from r_vector and z_vector using the 
\code{pretty} function. This can be overwritten by passing tick points in 
r_axis_ticks and z_axis_ticks

```{r}
four_grid = polar_grid(r_vector=syn_polar@polar$r_zscore, 
                       z_vector=-log10(syn_polar@pvalues$`LRT pvalue`), 
                       r_axis_ticks = NULL,
                       z_axis_ticks = NULL,
                       n_spokes = 4)

volcano3D(syn_polar,
          grid = four_grid, 
          label_rows = c("SLAMF6", "PARP16", "ITM2C"),
          label_size = 10,
          xy_aspectratio = 1,
          z_aspectratio = 0.9)
```

```{r}
three_grid = polar_grid(r_vector=NULL, 
                        z_vector=NULL, 
                        z_axis_ticks = c(0, 8, 16, 32),
                        r_axis_ticks = c(1, 2, 3),
                        n_spokes = 24)

radial_ggplot(polar = syn_polar,
              grid = three_grid,
              fc_cutoff = 0.1,
              marker_size = 2.3,
              legend_size = 10)

radial_ggplot(polar = syn_polar,
              fc_cutoff = 0.1,
              grid = three_grid, 
              label_rows = c("SLAMF6", "PARP16", "ITM2C"),
              marker_size = 2.3,
              legend_size = 10)
```



<!-- ## 2. Synovial Modular Data -->

<!-- We can then collapse this into a modular analysis using a list of gene sets. -->
<!-- In this example we have used the blood transcript modules curated by Li et. al. -->
<!-- in ['Li, S., Rouphael, N., Duraisingham, S., Romero-Steiner, S., Presnell, S., -->
<!-- Davis, C., ... & Kasturi, S. (2014). Molecular signatures of antibody responses -->
<!-- derived from a systems biology study of five human vaccines. Nature immunology, -->
<!-- 15(2), 195.'](https://www.nature.com/ni/journal/v15/n2/abs/ni.2789.html). The -->
<!-- pvalues were generated using  -->
<!-- [QuSAGE methodology](http://clip.med.yale.edu/qusage/). -->

<!-- ### Creating dep object -->

<!-- ```{r} -->
<!-- syn_mod_p_obj <- create_dep(sampledata = syn_metadata, -->
<!--                            contrast = "Pathotype", -->
<!--                            pvalues = syn_mod_pvalues, -->
<!--                            p_col_suffix = "p.value", -->
<!--                            padj_col_suffix = "q.value", -->
<!--                            fc_col_suffix = "logFC", -->
<!--                            multi_group_prefix = NULL, -->
<!--                            expression = syn_mod) -->
<!-- ``` -->


<!-- ### Volcano Plots -->

<!-- We can now investigate the comparisons between pathotypes using the -->
<!-- volcano_trio function. -->

<!-- ```{r, message = FALSE, fig.height = 2.8} -->
<!-- syn_mod_plots <- volcano_trio(dep = syn_mod_p_obj, -->
<!--                               label_rows = c("M156.0", "M37.2"),  -->
<!--                               shared_legend_size = 1,  -->
<!--                               sig_names = c("Not Sig",  -->
<!--                                             paste("Padj <", 0.05),  -->
<!--                                             paste("|FC| >", 1),  -->
<!--                                             paste("Padj <", 0.05,  -->
<!--                                                   "&\n|FC| >", 1))) -->

<!-- syn_mod_plots$All -->
<!-- ``` -->

<!-- ### Polar Plots -->

<!-- Next the coordinates and colours to map each variable, or row, in the -->
<!-- expression/pvalues data onto a polar coordinate system is calculated using the -->
<!-- polar_coords function: -->

<!-- ```{r} -->
<!-- syn_polar <- polar_coords(dep = syn_mod_p_obj, significance_cutoff = 0.01) -->

<!-- table(syn_polar@polar$sig) %>% -->
<!--   kable(col.names = c("Significance", "Frequency")) %>% -->
<!--   kable_styling(full_width = F) -->
<!-- ``` -->

<!-- These can be output to an interactive radar plot using radial_plotly. The -->
<!-- `labelRows` variable allows any markers of interest to be labelled. -->

<!-- ```{r, warning = FALSE, message = FALSE, out.height = "100%"} -->
<!-- radial_plotly(polar = syn_polar, -->
<!--               fc_cutoff = 0.3, -->
<!--               label_rows = c("M156.0", "M37.2")) %>% -->
<!--   layout(width = 650, height = 650) -->
<!-- ``` -->

<!-- Or a ggplot static image using radial_ggplot: -->

<!-- ```{r, warning = FALSE, fig.height = 5} -->
<!-- radial_ggplot(polar = syn_polar, -->
<!--               fc_cutoff = 0.1, -->
<!--               label_rows = c("M156.0", "M37.2"), -->
<!--               marker_size = 2.7, -->
<!--               label_size = 5, -->
<!--               axis_lab_size = 3, -->
<!--               axis_title_size = 5, -->
<!--               legend_size = 10) + -->
<!--   theme(legend.position = "right") -->
<!-- ``` -->

<!-- ### Boxplots -->

<!-- We can then interrogate any one specific variable as a boxplot, to investigate -->
<!-- these differences.  -->

<!-- ```{r, fig.height = 2.8, fig.width=5} -->
<!-- plot1 <- boxplot_trio(syn_mod_p_obj, -->
<!--                      value = "M156.0", -->
<!--                      test = "wilcox.test", -->
<!--                      levels_order = c("Lymphoid", "Myeloid", "Fibroid"), -->
<!--                      box_colours = c("blue", "red", "green3")) -->

<!-- plot2 <- boxplot_trio(syn_mod_p_obj, -->
<!--                      value = "M37.2", -->
<!--                      test = "wilcox.test", -->
<!--                      levels_order = c("Lymphoid", "Myeloid", "Fibroid"), -->
<!--                      box_colours = c("blue", "red", "green3")) -->

<!-- ggarrange(plot1, plot2) -->
<!-- ``` -->

<!-- --- -->

<!-- # Citation -->

<!-- If you use this package please cite as:  -->

<!-- > Lewis, Myles J., et al. "Molecular portraits of early rheumatoid arthritis  -->
<!-- identify clinical and treatment response phenotypes." Cell reports 28.9 (2019):  -->
<!-- 2455-2470. -->

<!-- or using: -->

<!-- ```{r} -->
<!-- citation("volcano3D") -->
<!-- ``` -->

