<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pvalues Generator • volcano3D</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Pvalues Generator">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-158679490-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-158679490-1');
</script>
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">volcano3D</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Vignette.html">Vignette</a>
    </li>
    <li>
      <a href="../articles/pvalues_generator.html">Creating Pvalues</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/KatrionaGoldmann/volcano3D">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="pvalues_generator_files/kePrint-0.0.1/kePrint.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Pvalues Generator</h1>
                        <h4 class="author">Katriona Goldmann</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/KatrionaGoldmann/volcano3D/blob/master/vignettes/pvalues_generator.rmd"><code>vignettes/pvalues_generator.rmd</code></a></small>
      <div class="hidden name"><code>pvalues_generator.rmd</code></div>

    </div>

    
    
<p><img src="../logo.png" align="right" alt="" width="200"></p>
<ul>
<li>
<a href="#introduction">Introduction</a>
<ul>
<li>
<a href="#getting-started">Getting Started</a>
<ul>
<li><a href="#install-from-cran">Install from CRAN</a></li>
<li><a href="#install-from-github">Install from Github</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#deseq-method">DESeq method</a></li>
</ul>
<hr>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This vignette provides an example for creating pvalues objects for the volcano3D pipeline using DESeq2 and limma.</p>
<p>This example consists of a case study from the PEAC rheumatoid arthritis project (Pathobiology of Early Arthritis Cohort). The methodology has been published in <a href="https://doi.org/10.1016/j.celrep.2019.07.091">‘Lewis, Myles J., et al. “Molecular portraits of early rheumatoid arthritis identify clinical and treatment response phenotypes.” Cell reports 28.9 (2019): 2455-2470.’ (DOI: 10.1016/j.celrep.2019.07.091)</a> with an interactive web tool available at <a href="https://peac.hpc.qmul.ac.uk">https://peac.hpc.qmul.ac.uk</a>.</p>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting Started</h2>
<div id="install-from-cran" class="section level3">
<h3 class="hasAnchor">
<a href="#install-from-cran" class="anchor"></a>Install from CRAN</h3>
<p><em>Not yet publicly available:</em></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"volcano3D"</span>)</a></code></pre></div>
</div>
<div id="install-from-github" class="section level3">
<h3 class="hasAnchor">
<a href="#install-from-github" class="anchor"></a>Install from Github</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(devtools)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"KatrionaGoldmann/volcano3D"</span>)</a></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(volcano3D)</a></code></pre></div>
<p>The sample data used in this vignette can be loaded from the <a href="https://github.com/KatrionaGoldmann/volcano3Ddata">volcano3Ddata package</a>. This is only possible after volcano3D is imported.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"volcano3Ddata"</span>)</a></code></pre></div>
<hr>
<p>To create a pvalues data frame we require:</p>
<ol style="list-style-type: decimal">
<li>the txi or expression data with columns representing different samples and rows representing different variables</li>
<li>The sample data which contains information for each sample in rows.</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(volcano3Ddata)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"syn_data"</span>)</a></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"syn_txi"</span>)</a></code></pre></div>
<p>Check the alignment.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="cf">if</span>(<span class="op">!</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/identical.html">identical</a></span>(syn_metadata<span class="op">$</span>ID, <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(syn_txi<span class="op">$</span>abundance))){</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">    <span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="st">"mis-aligned"</span>)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">}</a></code></pre></div>
<p>Make sure there are only three possible contrasts (i.e. three unique Pathotypes)</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="cf">if</span>(<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(syn_metadata<span class="op">$</span>Pathotype)) <span class="op">!=</span><span class="st"> </span><span class="dv">3</span>){</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">    <span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="st">"The number of unique pathotypes must qual 3"</span>)</a>
<a class="sourceLine" id="cb8-3" data-line-number="3">}</a></code></pre></div>
<p>Comparisons of interest:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">groups &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(syn_metadata<span class="op">$</span>Pathotype)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">contrasts &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(groups[<span class="dv">1</span>], groups[<span class="dv">2</span>], <span class="dt">sep=</span><span class="st">"-"</span>), </a>
<a class="sourceLine" id="cb9-3" data-line-number="3">               <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(groups[<span class="dv">2</span>], groups[<span class="dv">3</span>], <span class="dt">sep=</span><span class="st">"-"</span>), </a>
<a class="sourceLine" id="cb9-4" data-line-number="4">               <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(groups[<span class="dv">3</span>], groups[<span class="dv">1</span>], <span class="dt">sep=</span><span class="st">"-"</span>))</a>
<a class="sourceLine" id="cb9-5" data-line-number="5"></a>
<a class="sourceLine" id="cb9-6" data-line-number="6"><span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(contrasts, <span class="dt">collapse=</span><span class="st">", "</span>))</a></code></pre></div>
<pre><code>## [1] "Fibroid-Lymphoid, Lymphoid-Myeloid, Myeloid-Fibroid"</code></pre>
</div>
</div>
</div>
<div id="deseq-method" class="section level1">
<h1 class="hasAnchor">
<a href="#deseq-method" class="anchor"></a>DESeq method</h1>
<p>Use the DESeq package to calculate the differential expression between groups. The vignette for this package can be found <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">here</a></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(DESeq2)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2">dds =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html">DESeqDataSetFromTximport</a></span>(<span class="dt">txi =</span> syn_txi, </a>
<a class="sourceLine" id="cb11-3" data-line-number="3">                               <span class="dt">colData =</span> syn_metadata, </a>
<a class="sourceLine" id="cb11-4" data-line-number="4">                               <span class="dt">design =</span> <span class="op">~</span>Pathotype<span class="op">+</span>Batch<span class="op">+</span>Gender)</a>
<a class="sourceLine" id="cb11-5" data-line-number="5"></a>
<a class="sourceLine" id="cb11-6" data-line-number="6">dds_DE &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq</a></span>(dds)</a>
<a class="sourceLine" id="cb11-7" data-line-number="7">dds_LRT &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html">DESeq</a></span>(dds, <span class="dt">test =</span> <span class="st">"LRT"</span>, <span class="dt">reduced =</span> <span class="op">~</span>Batch<span class="op">+</span>Gender, <span class="dt">parallel =</span> <span class="ot">TRUE</span>) </a></code></pre></div>
<p>Output the results into a dataframe. First for each contrast:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">Pvals_DESeq_DE &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(contrasts, <span class="cf">function</span>(x){</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">    vars &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(x, <span class="dt">split=</span><span class="st">"-"</span>))</a>
<a class="sourceLine" id="cb12-3" data-line-number="3">    out &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span>(dds_DE, <span class="dt">contrast=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Pathotype"</span>, vars))</a>
<a class="sourceLine" id="cb12-4" data-line-number="4">    out &lt;-<span class="st"> </span>out[<span class="kw"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(syn_txi<span class="op">$</span>counts), <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(out)), ]</a>
<a class="sourceLine" id="cb12-5" data-line-number="5">    out &lt;-<span class="st"> </span>out[,<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"pvalue"</span>, <span class="st">"padj"</span>, <span class="st">"log2FoldChange"</span>)]</a>
<a class="sourceLine" id="cb12-6" data-line-number="6">})</a></code></pre></div>
<p>Then for a multi-group test. In this case we use the Likelihood ratio test.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">LRT &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/DESeq2/man/results.html">results</a></span>(dds_LRT, <span class="dt">parallel=</span><span class="ot">TRUE</span>)[, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"pvalue"</span>, <span class="st">"padj"</span>, <span class="st">"log2FoldChange"</span>)]</a>
<a class="sourceLine" id="cb13-2" data-line-number="2">LRT &lt;-<span class="st"> </span>LRT[<span class="kw"><a href="https://rdrr.io/r/base/match.html">match</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(syn_txi<span class="op">$</span>counts), <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(LRT)), ]</a></code></pre></div>
<p>Combine these into one pvalues object:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">Pvals_DESeq &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(Pvals_DESeq_DE[[<span class="dv">1</span>]], </a>
<a class="sourceLine" id="cb14-2" data-line-number="2">                     Pvals_DESeq_DE[[<span class="dv">2</span>]], </a>
<a class="sourceLine" id="cb14-3" data-line-number="3">                     Pvals_DESeq_DE[[<span class="dv">3</span>]], </a>
<a class="sourceLine" id="cb14-4" data-line-number="4">                     LRT)</a>
<a class="sourceLine" id="cb14-5" data-line-number="5"><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(Pvals_DESeq) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(contrasts, <span class="dt">each=</span><span class="dv">3</span>), </a>
<a class="sourceLine" id="cb14-6" data-line-number="6">                           <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(Pvals_DESeq_DE[[<span class="dv">1</span>]]), <span class="dv">3</span>)), </a>
<a class="sourceLine" id="cb14-7" data-line-number="7">                     <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"LRT"</span>, <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(LRT)))</a>
<a class="sourceLine" id="cb14-8" data-line-number="8"></a>
<a class="sourceLine" id="cb14-9" data-line-number="9">Pvals_DESeq &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(Pvals_DESeq[<span class="kw"><a href="https://rdrr.io/r/stats/complete.cases.html">complete.cases</a></span>(Pvals_DESeq), ], </a>
<a class="sourceLine" id="cb14-10" data-line-number="10">                          <span class="dt">check.names =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p>Now we can inspect:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(Pvals_DESeq)</a></code></pre></div>
<table class="table table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Fibroid-Lymphoid pvalue
</th>
<th style="text-align:right;">
Fibroid-Lymphoid padj
</th>
<th style="text-align:right;">
Fibroid-Lymphoid log2FoldChange
</th>
<th style="text-align:right;">
Lymphoid-Myeloid pvalue
</th>
<th style="text-align:right;">
Lymphoid-Myeloid padj
</th>
<th style="text-align:right;">
Lymphoid-Myeloid log2FoldChange
</th>
<th style="text-align:right;">
Myeloid-Fibroid pvalue
</th>
<th style="text-align:right;">
Myeloid-Fibroid padj
</th>
<th style="text-align:right;">
Myeloid-Fibroid log2FoldChange
</th>
<th style="text-align:right;">
LRT pvalue
</th>
<th style="text-align:right;">
LRT padj
</th>
<th style="text-align:right;">
LRT log2FoldChange
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
A2M
</td>
<td style="text-align:right;">
0.7775361
</td>
<td style="text-align:right;">
0.8851935
</td>
<td style="text-align:right;">
-0.0478326
</td>
<td style="text-align:right;">
0.1563045
</td>
<td style="text-align:right;">
0.3384292
</td>
<td style="text-align:right;">
-0.2202768
</td>
<td style="text-align:right;">
0.1694654
</td>
<td style="text-align:right;">
0.4916679
</td>
<td style="text-align:right;">
0.2681094
</td>
<td style="text-align:right;">
0.2739571
</td>
<td style="text-align:right;">
0.4353436
</td>
<td style="text-align:right;">
0.0527816
</td>
</tr>
<tr>
<td style="text-align:left;">
A2ML1
</td>
<td style="text-align:right;">
0.0002106
</td>
<td style="text-align:right;">
0.0011118
</td>
<td style="text-align:right;">
1.6854024
</td>
<td style="text-align:right;">
0.0000030
</td>
<td style="text-align:right;">
0.0000793
</td>
<td style="text-align:right;">
-1.9299687
</td>
<td style="text-align:right;">
0.6348844
</td>
<td style="text-align:right;">
0.8441036
</td>
<td style="text-align:right;">
0.2445664
</td>
<td style="text-align:right;">
0.0000022
</td>
<td style="text-align:right;">
0.0000202
</td>
<td style="text-align:right;">
0.3983423
</td>
</tr>
<tr>
<td style="text-align:left;">
A4GALT
</td>
<td style="text-align:right;">
0.0000000
</td>
<td style="text-align:right;">
0.0000000
</td>
<td style="text-align:right;">
1.0248506
</td>
<td style="text-align:right;">
0.0000192
</td>
<td style="text-align:right;">
0.0003539
</td>
<td style="text-align:right;">
-0.5646665
</td>
<td style="text-align:right;">
0.0055222
</td>
<td style="text-align:right;">
0.1009600
</td>
<td style="text-align:right;">
-0.4601841
</td>
<td style="text-align:right;">
0.0000000
</td>
<td style="text-align:right;">
0.0000000
</td>
<td style="text-align:right;">
0.2121462
</td>
</tr>
<tr>
<td style="text-align:left;">
A4GNT
</td>
<td style="text-align:right;">
0.0238302
</td>
<td style="text-align:right;">
0.0610621
</td>
<td style="text-align:right;">
-1.1527859
</td>
<td style="text-align:right;">
0.6691773
</td>
<td style="text-align:right;">
0.8180681
</td>
<td style="text-align:right;">
-0.1857435
</td>
<td style="text-align:right;">
0.0197330
</td>
<td style="text-align:right;">
0.1893057
</td>
<td style="text-align:right;">
1.3385294
</td>
<td style="text-align:right;">
0.0732984
</td>
<td style="text-align:right;">
0.1583021
</td>
<td style="text-align:right;">
-0.1939520
</td>
</tr>
<tr>
<td style="text-align:left;">
AAAS
</td>
<td style="text-align:right;">
0.8574281
</td>
<td style="text-align:right;">
0.9383721
</td>
<td style="text-align:right;">
-0.0245296
</td>
<td style="text-align:right;">
0.6608837
</td>
<td style="text-align:right;">
0.8124146
</td>
<td style="text-align:right;">
0.0548697
</td>
<td style="text-align:right;">
0.8469982
</td>
<td style="text-align:right;">
0.9435306
</td>
<td style="text-align:right;">
-0.0303402
</td>
<td style="text-align:right;">
0.9075226
</td>
<td style="text-align:right;">
0.9711813
</td>
<td style="text-align:right;">
0.3523114
</td>
</tr>
<tr>
<td style="text-align:left;">
AACS
</td>
<td style="text-align:right;">
0.2815302
</td>
<td style="text-align:right;">
0.4322015
</td>
<td style="text-align:right;">
0.1729592
</td>
<td style="text-align:right;">
0.0630182
</td>
<td style="text-align:right;">
0.1821943
</td>
<td style="text-align:right;">
-0.2732040
</td>
<td style="text-align:right;">
0.5874802
</td>
<td style="text-align:right;">
0.8169632
</td>
<td style="text-align:right;">
0.1002448
</td>
<td style="text-align:right;">
0.1472059
</td>
<td style="text-align:right;">
0.2732938
</td>
<td style="text-align:right;">
-0.2881377
</td>
</tr>
</tbody>
</table>
<!-- # Limma method --><!-- Use the Limma and edgeR package to calculate the differential expression  --><!-- between groups. The vignette for this package can be found [here](https://www.bioconductor.org/packages/devel/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html) --><!-- ```{r} --><!-- library(limma) --><!-- library(edgeR) --><!-- load("~/Downloads/syn_txi.rdata") --><!-- ``` --><!-- Define the design for Limma --><!-- ```{r, eval=FALSE} --><!-- design.df <- model.matrix(~ 0 + Pathotype + Batch + Gender, data=syn_metadata)  --><!-- colnames(design.df) <- gsub("[[:punct:]]|factor|syn_metadata|Pathotype|Batch",  --><!--                             "", colnames(design.df)) --><!-- contrast.matrix <- makeContrasts(contrasts=c(contrasts, "B", "GenderB"),  --><!--                                  levels = c(groups,"B", "GenderB") ) --><!-- ``` --><!-- Remove the batch effect: --><!-- ```{r} --><!-- exp <- data.frame(apply(syn_rld, 2, as.numeric)) --><!-- rownames(exp) <- make.names(rownames(syn_rld), unique=T) --><!-- ``` --><!-- Calculate the differential expression --><!-- ```{r, eval=FALSE} --><!-- fit <- lmFit(exp, design.df) --><!-- fit2  <- contrasts.fit(fit, contrasts=contrast.matrix) --><!-- fit3  <- eBayes(fit2, trend = TRUE) --><!-- ``` --><!-- Output the p-values first for individual comparisons: --><!-- ```{r, eval=FALSE} --><!-- Pvals_limma_DE <- lapply(contrasts, function(x){ --><!--     out <- topTable(fit3,  --><!--              coef=which(colnames(fit3$coefficients) == x),  --><!--              number=Inf,  --><!--              sort.by="P")[,c("P.Value", "adj.P.Val", "logFC")] --><!--     out <- out[match(rownames(exp), rownames(out)), ] --><!--     colnames(out) <- c("pvalue", "padj", "log2FoldChange") --><!--     out --><!-- }) --><!-- ``` --><!-- Then for an empirical Bayes multi-group test: --><!-- ```{r, eval=FALSE} --><!-- eB <- topTable(fit3, number=Inf,  --><!--                 coef=which(colnames(fit3$coefficients) %in% contrasts)) --><!-- eB <- eB[match(rownames(exp), rownames(eB)), ] --><!-- eB <- eB[, ! colnames(eB) %in% c("AveExpr", "F")] --><!-- colnames(eB) = c(contrasts,  "pvalue", "padj" ) --><!-- ``` --><!-- Combined into one object --><!-- ```{r, eval=FALSE} --><!-- Pvals_limma <- cbind(Pvals_limma_DE[[1]],  --><!--                      Pvals_limma_DE[[2]],  --><!--                      Pvals_limma_DE[[3]],  --><!--                      "eBayes" = eB) --><!-- colnames(Pvals_limma) <- c(paste(rep(contrasts, each=3),  --><!--                            rep(colnames(Pvals_limma_DE[[1]]), 3)),  --><!--                      paste("eBayes", colnames(eB))) --><!-- Pvals_limma <- data.frame(Pvals_limma[complete.cases(Pvals_limma), ],  --><!--                           check.names = FALSE) --><!-- ``` --><!-- Now we can inspect: --><!-- ```{r, eval=FALSE} --><!-- head(Pvals_limma) --><!-- ``` --><!-- ```{r, echo=FALSE} --><!-- Pvals_limma[, 1:(ncol(Pvals_limma)-1)] = format( --><!--     Pvals_limma[, 1:(ncol(Pvals_limma)-1)], digits = 3) --><!-- head(Pvals_limma) %>% --><!--     kable() %>% --><!--     kable_styling(font_size=8.7) --><!-- ``` --><!-- These objects can be used as pvalues dataframes in downstream analysis with the --><!-- volcano3D package.  -->
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#introduction">Introduction</a><ul class="nav nav-pills nav-stacked">
<li><a href="#getting-started">Getting Started</a></li>
      </ul>
</li>
      <li><a href="#deseq-method">DESeq method</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/KatrionaGoldmann">Katriona Goldmann</a>, <a href="https://www.qmul.ac.uk/whri/people/academic-staff/items/lewismyles.html">Myles Lewis</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
